### ISSUE TITLE
Extend the H-axis reduce split on TILE inputs for ttnn.mean / ttnn.sum

### ISSUE THREAD

**Issue by ign-msati** [tenstorrent/tt-metal#52922]
### Is your feature request related to a problem? Please describe

<!DOCTYPE html><p cid="n6509" mdtype="paragraph" class="md-end-block md-p md-focus" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin-top: 0px; margin-bottom: 1.5rem; overflow-wrap: break-word; white-space: pre-wrap; position: relative; color: rgb(184, 191, 198); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, &quot;SF Pro&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span md-inline="plain" class="md-plain md-expand" style="box-sizing: border-box;">An H-axis reduce parallelizes only over output tile columns — </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">num_cols = NC * Wt</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> (</span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">reduce_op_multi_core_h_program_factory.cpp:92</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">) — so a narrow-</span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">W</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> tensor starves the grid no matter how tall it is. </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">(1, 1, 51264, 32)</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> gives </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">NC * Wt == 1</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">: one work unit, one core, 51264 rows serialized, 63 cores idle.</span></p><p cid="n6510" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin-top: 0px; margin-bottom: 1.5rem; overflow-wrap: break-word; white-space: pre-wrap; position: relative; color: rgb(184, 191, 198); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, &quot;SF Pro&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span md-inline="strong" class="md-pair-s " style="box-sizing: border-box;"><strong style="box-sizing: border-box; font-weight: bold; color: rgb(222, 222, 222);"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">#50393 fixes this, but only for ROW_MAJOR inputs.</span></strong></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> It adds an H-axis split that segments the reduce axis into </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">S</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> slices and collapses them in a second stage. The split is gated on the row-major dense path, so TILE inputs keep the one-core behaviour.</span></p><p cid="n6511" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin-top: 0px; margin-bottom: 1.5rem; overflow-wrap: break-word; white-space: pre-wrap; position: relative; color: rgb(184, 191, 198); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, &quot;SF Pro&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Measured with </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">ttnn.avg_pool2d</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> as a global pool over </span><span md-inline="code" spellcheck="false" class="md-pair-s" style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: Monaco, Consolas, &quot;Andale Mono&quot;, &quot;DejaVu Sans Mono&quot;, monospace; text-align: left; vertical-align: initial; font-size: 0.875em; background: rgba(0, 0, 0, 0.05); padding: 2px 5px;">(1, 1, 51264, 32)</code></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> bfloat16 — same op, same data, only the input layout differs:</span></p><figure class="md-table-fig table-figure" cid="n6512" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; color: rgb(184, 191, 198); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, &quot;SF Pro&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
input layout | main | with #50393
-- | -- | --
ROW_MAJOR | 1016.2 us | 241.7 us
TILE | 914.9 us | 916.8 us (unchanged)

</figure>

### Describe the solution you'd like

Support the H-axis split for **`ttnn.mean` and `ttnn.sum` on TILE inputs**, reusing #50393's two-stage shape: stage 1 reduces per-slice into FP32 partials, stage 2 collapses the slice axis and applies the user scaler. Scoping it to mean/sum matches #50393's own scope, so the two changes stay consistent.

**The same mechanism also applies to ttnn.min / ttnn.max**. This is worth noting because the row-major path in #50393 can never reach them, making the split on the tiled side the only version where this optimization applies

### Additional context

```
import pytest
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

pytestmark = [pytest.mark.use_module_device]

TEST_PADDING_VALUE = -42


@pytest.mark.parametrize("dtype", [ttnn.bfloat16, ttnn.float32])
@pytest.mark.parametrize(
    "input_layout",
    [ttnn.ROW_MAJOR_LAYOUT, ttnn.TILE_LAYOUT],
    ids=["in_rm", "in_tile"],
)
@pytest.mark.parametrize(
    "output_layout",
    [ttnn.ROW_MAJOR_LAYOUT, ttnn.TILE_LAYOUT],
    ids=["out_rm", "out_tile"],
)
# (N, 1, H*W, C) — reduced over the H*W axis.
@pytest.mark.parametrize(
    "shape",
    [
        (1, 1, 51264, 32),
        (1, 1, 12832, 64),
        (1, 1, 3216, 128),
        (1, 1, 808, 256),
    ],
)
def test_mean(device, dtype, input_layout, output_layout, shape):
    """Global mean over the flattened spatial axis."""
    _, _, hw, _ = shape

    torch.manual_seed(0)
    torch_input = torch.rand(shape, dtype=torch.float32 if dtype == ttnn.float32 else torch.bfloat16)
    torch_ref = torch.mean(torch_input.float(), dim=2, keepdim=True).to(torch_input.dtype)

    tt_input = ttnn.from_torch(torch_input, dtype=dtype, layout=input_layout, device=device)
    assert tt_input.layout == input_layout
    if input_layout == ttnn.TILE_LAYOUT:
        # Non-aligned H*W (3216, 808) pads to a tile row; poison pad so garbage can't pass.
        tt_input = ttnn.fill_implicit_tile_padding(tt_input, TEST_PADDING_VALUE)

    tt_output = ttnn.mean(tt_input, dim=-2, keepdim=True, output_layout=output_layout)
    assert tt_output.layout == output_layout
    output = ttnn.to_torch(tt_output)

    assert output.shape == torch_ref.shape
    # Tall bf16 means sit near 0.5 with tiny channel variance, so PCC is unstable there
    # even when the values themselves are close.
    if dtype == ttnn.float32:
        pcc = 0.999
    else:
         pcc = 0.70 if hw >= 12832 else 0.97
    assert_with_pcc(torch_ref, output, pcc)
```

<!DOCTYPE html>
Shape | Dtype | In layout | Out layout | Time (µs) https://github.com/tenstorrent/tt-metal/pull/50393' | Cores | H-slices
-- | -- | -- | -- | -- | -- | --
1,1,51264,32 | bfloat16 | ROW_MAJOR | ROW_MAJOR | 241.7 | 64/64 | 64
1,1,51264,32 | bfloat16 | ROW_MAJOR | TILE | 249.6 | 64/64 | 64
1,1,51264,32 | bfloat16 | TILE | ROW_MAJOR | 915.6 | 1/64 | 1
1,1,51264,32 | bfloat16 | TILE | TILE | 916.8 | 1/64 | 1
1,1,51264,32 | float32 | ROW_MAJOR | ROW_MAJOR | 200.7 | 64/64 | 64
1,1,51264,32 | float32 | ROW_MAJOR | TILE | 211.4 | 64/64 | 64
1,1,51264,32 | float32 | TILE | ROW_MAJOR | 1120.4 | 1/64 | 1
1,1,51264,32 | float32 | TILE | TILE | 1122.8 | 1/64 | 1
